code/modules/station_goals/bluespace_tap.dm

```diff
@@ -0,0 +1,359 @@
+		occupied += get_step(src,direct)
+	occupied += locate(x+1,y-2,z)
+	occupied += locate(x-1,y-2,z)
+
+	for(var/T in occupied)
+		var/obj/structure/filler/F = new(T)
+		F.parent = src
+		fillers += F
+	if(!powernet)
+		connect_to_network()
+
+/obj/machinery/power/bluespace_tap/Destroy()
+	for(var/V in fillers)
+		var/obj/structure/filler/filler = V
+		filler.parent = null
+		qdel(filler)
+	fillers.Cut()
+	. = ..()
+
+/obj/machinery/power/bluespace_tap/proc/increase_level()
+	if(input_level < max_level)
+		input_level++
+
+/obj/machinery/power/bluespace_tap/proc/decrease_level()
+	if(input_level > 0)
+		input_level--
+
+
+/obj/machinery/power/bluespace_tap/proc/set_level(desired_level)
+	if(desired_level < 0)
+		return
+	if(desired_level > max_level)
+		return
+
+	input_level = desired_level
+
+
+
+//stuff that happens regularily, ie power use and point generation
+/obj/machinery/power/bluespace_tap/process()
+	if(input_level == 0)
+		actual_power_usage = 0
+		return
+	actual_power_usage = (base_value ** input_level) * active_power_usage	//each level takes one order of magnitude more power than the previous one
+	if(surplus() < actual_power_usage)
+		decrease_level()	//turn down a level when lacking power, don't want to suck the powernet dry
+		return
+	else
+		add_load(actual_power_usage)
+		var/points_to_add = (input_level + emagged) * base_points
+		points += points_to_add	//point generation, emagging gets you 'free' points at the cost of higher anomaly chance
+		total_points += points_to_add
+		if(prob(input_level - safe_levels + (emagged * 5)))	//at dangerous levels, start doing freaky shit. prob with values less than 0 treat it as 0, so only occurs if input level > 7
+			event_announcement.Announce("Unexpected power spike during Bluespace Harvester Operation. Extra-dimensional intruder alert. Expected location: [get_area(src).name].", "Bluespace Harvester Malfunction")
+			if(!emagged)
+				input_level = 0	//as hilarious as it would be for the tap to spawn in even more nasties because you can't get to it to turn it off, that might be too much for now. Unless sabotage is involved
+			for(var/i = 1, i <= rand(1, 3), i++)	//freaky shit here, 1-3 freaky portals
+				var/turf/location = locate(x + rand(-5,5), y + rand(-5,5), z)
+				var/obj/structure/spawner/nether/bluespace_tap/portal = new(src)
+				portal.forceMove(location)
+
+
+
+/obj/machinery/power/bluespace_tap/ui_data(mob/user, ui_key = "main", datum/topic_state/state = default_state)
+	var/data[0]
+
+	data["level"] = input_level
+	data["points"] = points
+	data["total_points"] = total_points
+	data["power_use"] = actual_power_usage
+	data["available_power"] = surplus()
+	data["max_level"] = max_level
+	data["emagged"] = emagged
+
+	var/list/listed_items = list()//a list of lists, each inner list equals a datum
+	for(var/key = 1 to length(product_list))
+		var/datum/data/bluespace_tap_product/A = product_list[key]
+		listed_items.Add(list(list(
```

Comment:
list of lists of lists? Might have put a list( to much here

Reply:
It's actually just a list of lists. if you do list.Add and pass it a list, it will add every element of that list to the original list. The documentation says 

> Appends the specified items to the list. If an argument is itself a list, each item in the list will be added. 

So I need to pack it in a second list. You can see the same code in vending.dm where I took it from.

Reply:
Ah just a weird byond thing then. Makes sense...